rust/src/pgsql/pgsql.rs

```diff
@@ -356,9 +376,22 @@ impl PgsqlState {
                         _needed,
                         needed_estimation
                     );
+                    SCLogDebug!("PgsqlEvent::TruncatedData");
+                    self.set_event(PgsqlEvent::TruncatedData);
                     return AppLayerResult::incomplete(consumed as u32, needed_estimation as u32);
                 }
+                Err(Err::Error(Error{code:ErrorKind::Verify, ..})) => {
+                    SCLogDebug!("PgsqlEvent::InvalidLength");
+                    self.set_event(PgsqlEvent::InvalidLength);
+                    AppLayerResult::err();
+                }
+                Err(Err::Error(Error{code:ErrorKind::Switch, ..})) => {
+                    self.set_event(PgsqlEvent::MalformedData);
+                    SCLogDebug!("PgsqlEvent::MalformedData");
+                    return AppLayerResult::ok();
```

Comment:
I wonder if it is possible to organize the code a bit differently :
1. get the PDU (or error if there is an invalid length)
2. parse the PDU (and if there is an error, still return `AppLayerResult::ok()` for next PDUs)

Reply:
(I think that the two tickets should have their own commit)

Reply:
Roger that on the commit per ticket part!
wrt to the re-structuring, I'm trying to visualize how to do that with pgsql, as the protocol only has pseudo-headers, and some PDUs may come with pure data, for which information on how to parse them (including length) would be present in a previous message, only...

Reply:
> and some PDUs may come with pure data

Which ones ?
It looked to me that it is always one byte opcode, and 4 bytes length of PDU...

Reply:
For some reason, I didn't see this question before T__T
When the pgsql backend is answering to a query request, for instance, a single PGSQL message may span over several PDUs, and the intermediate PGSQL messages don't come with any header info.
Taking https://github.com/OISF/suricata-verify/blob/master/tests/pgsql/pgsql-5000-query-results/input.pcap as an example, in packet no. 32, the second to last PGSQL message has an informed length of 482 bytes, but the PGSQL PDU only has 132 bytes, so the rest of the message spills onto the next PGSQL PDU, and for that one we don't have info on how to read it.
The info is actually on the same second-to-last message, which indicates that the next data will be a `Data` column with 360 bytes, but this is only present in the previous message. The last message starts in the middle of a rule signature, which is part of the `Data` result, which actually started on the previous message.
 I guess... this could become some transaction info that could be used for such cases?

Reply:
On packet 32, I see the second to last PGSQL message has an informed length of 616 bytes (0x268)

<img width="1904" alt="Capture d’écran 2023-11-16 à 17 27 35" src="https://github.com/OISF/suricata/assets/35799796/61970c30-5375-4f03-a519-40d31a623c3a">

Reply:
To me, it's 665:
![image](https://github.com/OISF/suricata/assets/16976857/3994525e-7207-4eff-a4ce-9baad82dff79)
But the overall thing is that in this case, the very last PGSQL message for this packet won't have any header info for us to process.

Reply:
With Wireshark Version 4.0.10 I do not have the last postgresql in packet 32 but I see it well reassembled in packet 34

Reply:
Thanks, I'll update mine, and maybe peak at how they handle this, I guess.

Reply:
I have finally updated my Wireshark. I managed to get version Wireshark 4.0.6, but I still see errors starting with packet 32... 🤔

Reply:
Still struggling a bit to figure out how to approach this bit for the PDU part. As I think of this, wondering if the introduction of the UnknownMessageType could be enough to handle this?
Introduced in 1ac5d97259994f73e645333 and 0fc9ade7a9f672560a276b

Reply:
> I wonder if it is possible to organize the code a bit differently :
> 
> 1. get the PDU (or error if there is an invalid length)
> 2. parse the PDU (and if there is an error, still return `AppLayerResult::ok()` for next PDUs)

Added this to a ticket, as it doesn't look trivial, still, for now https://redmine.openinfosecfoundation.org/issues/6654)

I think that part of this is somewhat addressed by the implementation of the `UnknownMessageTypes` - although not entirely, as we're not changing how the PDU is processed, just reducing the situations when the parser returns an error.

Reply:
> wondering if the introduction of the UnknownMessageType could be enough to handle this?

Yes but not sure it is the best way to have duplicated code to get the length everywhere...

But I guess the functionality (return AppLayerResult::ok() so as to parse next PDUs) can be achieved even if the code is duplicated...

Reply:
Fair point, I tried to reduce that with the length function, but there's certainly a lot of room for improvement...